services/statestorage/service.go

```diff
@@ -121,13 +176,38 @@ func (s *service) ReadKeys(ctx context.Context, input *services.ReadKeysInput) (
 	return output, nil
 }
 
+func (s *service) _readStateKey(height primitives.BlockHeight, contract primitives.ContractName, key string) (*protocol.StateRecord, bool, error) {
+	for i := len(s.incCache) - 1; i >= 0; i-- {
+		if s.incCache[i].height > height {
+			continue
+		}
+		if record, exists := s.incCache[i].diff[contract][key]; exists {
+			return record, !isZeroValue(record.Value()), nil // cached state increments must include zero values
+		}
+	}
+	return s.persistence.Read(contract, key)
+}
+
+func (s *service) _readStateHash(height primitives.BlockHeight) (primitives.MerkleSha256, error) {
+	persistedHeight, _, persistedHash, err := s.persistence.ReadMetadata()
+	if err != nil {
+		return nil, err
+	}
+	if height == persistedHeight {
+		return persistedHash, nil
+	}
+
+	cacheIdx := height - persistedHeight - 1
+	return s.incCache[cacheIdx].merkleRoot, nil
+}
+
 func (s *service) GetStateStorageBlockHeight(ctx context.Context, input *services.GetStateStorageBlockHeightInput) (*services.GetStateStorageBlockHeightOutput, error) {
 	s.mutex.RLock()
 	defer s.mutex.RUnlock()
 
 	result := &services.GetStateStorageBlockHeightOutput{
-		LastCommittedBlockHeight:    s.lastCommittedBlockHeader.BlockHeight(),
-		LastCommittedBlockTimestamp: s.lastCommittedBlockHeader.Timestamp(),
+		LastCommittedBlockHeight:    s.height,
```

Comment:
i prefer keeing lastCommited header instead of height and ts seperately
maybe even keep one copy of each lastCommitted for each history  diff.